core/src/main/scala/org/bitcoins/core/protocol/script/ScriptWitness.scala

```diff
@@ -208,3 +202,213 @@ object ScriptWitness extends Factory[ScriptWitness] {
     }
   }
 }
+
+sealed trait TaprootWitness extends ScriptWitness {
+  override def bytes: ByteVector = RawScriptWitnessParser.write(this)
+}
+
+object TaprootWitness {
+
+  def fromStack(stack: Vector[ByteVector]): TaprootWitness = {
+    if (stack.length == 1) TaprootKeyPath.fromStack(stack)
+    else TaprootScriptPath(stack)
+  }
+}
+
+/** Spending a taproot output via the key path spend */
+case class TaprootKeyPath(
+    signature: SchnorrDigitalSignature,
+    hashType: HashType)
+    extends TaprootWitness {
+  override val stack: Vector[ByteVector] = Vector(signature.bytes)
+}
+
+object TaprootKeyPath {
+
+  def fromStack(vec: Vector[ByteVector]): TaprootKeyPath = {
+    require(
+      vec.length == 1,
+      s"Taproot keypath can only have one stack element, got=${vec.length}")
+
+    val sigBytes = vec.head
+
+    val keyPath = if (sigBytes.length == 64) {
+      //means SIGHASH_ALL is implicitly encoded
+      //see: https://github.com/bitcoin/bips/blob/master/bip-0341.mediawiki#Common_signature_message
+      val sig = SchnorrDigitalSignature.fromBytes(sigBytes)
+      TaprootKeyPath(sig, HashType.sigHashAll)
+    } else if (sigBytes.length == 65) {
+      val sig = SchnorrDigitalSignature.fromBytes(sigBytes.dropRight(1))
+      val hashType = HashType.fromByte(sigBytes.last)
+      TaprootKeyPath(sig, hashType)
+    } else {
+      sys.error(
+        s"Unknown sig bytes length, should be 64 or 65, got=${sigBytes.length}")
+    }
+
+    keyPath
+  }
+
+  def isValid(stack: Vector[ByteVector]): Boolean = {
+    stack.length == 1 && (stack.length == 64 || stack.head.length == 65)
+  }
+}
+
+/** Spending a taproot output via the script path */
+case class TaprootScriptPath(stack: Vector[ByteVector]) extends TaprootWitness {
+  require(TaprootScriptPath.isValid(stack),
+          s"Invalid witness stack for TaprootScriptPath, got=$stack")
+
+  val controlBlock: ControlBlock = {
+    if (stack.headOption.map(_.head) == TaprootScriptPath.annexOpt) {
+      //If there are at least two witness elements, and the first byte of the last element is 0x50[4],
+      // this last element is called annex a[5] and is removed from the witness stack.
+      // The annex (or the lack of thereof) is always covered by the signature and contributes to transaction weight,
+      // but is otherwise ignored during taproot validation.
+      //see: https://github.com/bitcoin/bips/blob/master/bip-0341.mediawiki#script-validation-rules
+      ControlBlock.fromBytes(stack(1))
+    } else {
+      ControlBlock.fromBytes(stack.head)
+    }
+  }
+
+  //  require(
+  //    controlBlock.length >= 33 && controlBlock.length < 33 + (32 * 128),
+  //    s"Control block length is wrong, got=${controlBlock.length}, max=${33 + (32 * 128)}")
```

Comment:
should this be brought back?

Reply:
this is now checked in the `TaprootScriptPath.isValid()` method now.